biomass/GEOCARBON/convert.py

```diff
@@ -0,0 +1,99 @@
+import os
+import time
+
+import cftime
+import numpy as np
+import xarray as xr
+
+
+def CoarsenDataset(filename, res=0.5, ntile=None):
+    """Coarsens the source dataset by using xarray and dask to avoid large memory.
+
+    Parameters
+    ----------
+    filename : str
+        the netcdf file to coarsen
+    outfile : str
+        the name of the file for the coarsened output
+    res : float, optional
+        the approximate resolution of the coarsen output in degrees
+    ntile : int, optional the number of coarse cells to process at a
+        time, routine will process a ntile x ntile subgrid. Increase
+        to run faster, decrease to reduce the peak memory.
+
+    """
+    with xr.open_dataset(filename) as ds:
+        ds = ds.rename({"latitude": "lat", "longitude": "lon"})
+        res0 = float(ds.lon.diff(dim="lon").mean())
+        n = int(round(res / res0))
+        if ntile is not None:
+            ds = ds.chunk({"lat": ntile * n, "lon": ntile * n})
+        c = ds.coarsen({"lat": n, "lon": n}, boundary="pad").mean()
+    return c
+
+
+if __name__ == "__main__":
+
+    # If local file does not exist, download
+    remote_source = "https://www.bgc-jena.mpg.de/geodb/projects/DataDnld.php"
+    local_source = "Forest_Aboveground_Biomassv3.nc"
+    download_stamp = time.strftime(
+        "%Y-%m-%d", time.localtime(os.path.getmtime(local_source))
+    )
+    generate_stamp = time.strftime("%Y-%m-%d")
+
+    # Coarsen
+    ds = CoarsenDataset(local_source, res=0.5)
+
+    # For some unknown reason, we need to re-create the biomass
+    # dataarray such that the encoding works out properly. I was
+    # having the problem that unit conversions didn't fail, but
+    # neither did they change the magnitudes of the data.
+    ds["biomass"] = xr.DataArray(
+        np.ma.masked_invalid(
+            ds["Forest_Aboveground_Biomass_v3"]
+            .to_numpy()
+            .reshape((1,) + ds["Forest_Aboveground_Biomass_v3"].shape)
+        ),
+        dims=("time", "lat", "lon"),
+    )
+    ds["biomass"].attrs = {"long_name": "above_ground_biomass", "units": "Mg ha-1"}
+    ds = ds.drop_vars("Forest_Aboveground_Biomass_v3")
+
+    # Having trouble with the time so let's just re-encode it
+    ds["time"] = [cftime.DatetimeNoLeap(2015, 6, 1)]
+    tb = np.asarray(
+        [
+            [cftime.DatetimeNoLeap(2000, 1, 1), cftime.DatetimeNoLeap(2010, 1, 1)],
+        ]
+    )
+    ds["time_bnds"] = xr.DataArray(tb, dims=("time", "nv"))
+    ds["time"].encoding["units"] = "days since 2000-01-01"
+    ds["time"].attrs["bounds"] = "time_bnds"
+
+    ds = ds.sortby(list(ds.sizes.keys()))
+    ds.attrs = {}
+    ds.attrs["title"] = (
+        "GEOCARBON: Towards an Operational Global Carbon Observing System"
+    )
+    ds.attrs["version"] = "3"
+    ds.attrs["institution"] = "GEOCARBON (European FP7 project)"
+    ds.attrs["source"] = (
```

Comment:
This isn't what they mean by source actually. See here: "The method of production of the original data. If it was model-generated, source should name the model and its version, as specifically as could be useful. If it is observational, source should characterize it (e.g., "surface observation" or "radiosonde")."

Reply:
My issue is that I see very little information about this dataset included in the record. I am not sure of what it is comprised. Here is the header of the file you download from them:

```
netcdf Forest_Aboveground_Biomassv3 {
dimensions:
	longitude = 36000 ;
	latitude = 14002 ;
variables:
	double longitude(longitude) ;
		longitude:standard_name = "longitude" ;
		longitude:long_name = "longitude" ;
		longitude:units = "degrees_east" ;
		longitude:axis = "X" ;
	double latitude(latitude) ;
		latitude:standard_name = "latitude" ;
		latitude:long_name = "latitude" ;
		latitude:units = "degrees_north" ;
		latitude:axis = "Y" ;
	float Forest_Aboveground_Biomass_v3(latitude, longitude) ;
		Forest_Aboveground_Biomass_v3:long_name = "Forest_Aboveground_Biomass_v3" ;
		Forest_Aboveground_Biomass_v3:_FillValue = -3.4e+38f ;
		Forest_Aboveground_Biomass_v3:missing_value = -3.4e+38f ;
		Forest_Aboveground_Biomass_v3:projection = "+proj=longlat +datum=WGS84 +no_defs +ellps=WGS84 +towgs84=0,0,0" ;
		Forest_Aboveground_Biomass_v3:projection_format = "PROJ.4" ;
		Forest_Aboveground_Biomass_v3:min = 0. ;
		Forest_Aboveground_Biomass_v3:max = 611.399963378906 ;

// global attributes:
		:CDI = "Climate Data Interface version 1.6.9 (http://mpimet.mpg.de/cdi)" ;
		:history = "Wed Sep 09 15:23:36 2015: cdo -f nc4c -z zip copy Forest_Aboveground_Biomass_v3.nc Forest_Aboveground_Biomass_v3_2.nc" ;
		:Conventions = "CF-1.4" ;
		:created_by = "R, packages ncdf and raster (version 2.2-31)" ;
		:date = "2015-08-31 14:39:27" ;
		:CDO = "Climate Data Operators version 1.6.9 (http://mpimet.mpg.de/cdo)" ;
}
```

I found this publication: https://onlinelibrary.wiley.com/doi/abs/10.1111/gcb.13139 but it is for a Pan-tropical dataset and GEOCARBON is global.

I found this ppt slide deck: https://pdfs.semanticscholar.org/4868/c42005447915bda9617853408cac4f0df27b.pdf when I google 'GEOCARBON' but it isn't exactly a solid reference or contain any.

They point to this website: https://www.geocarbon.net/ but I didn't find any helpful information.

I am honestly not sure what to say.

Reply:
I think this might have what you need for some references: https://datacore-gn.unepgrid.ch/geonetwork/srv/api/records/5e695176-266d-4697-bc06-c2d9196845b4